File: src/main/java/com/atrilos/twoPointers/ThreeSum.java
package com.atrilos.twoPointers;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

/**
 * Given an integer array nums, return all the triplets [nums[i], nums[j], nums[k]] such that i != j, i != k, and j != k, and nums[i] + nums[j] + nums[k] == 0.
 * <p>
 * Notice that the solution set must not contain duplicate triplets.
 * <p>
 * <p>
 * <p>
 * Example 1:
 * <p>
 * Input: nums = [-1,0,1,2,-1,-4]
 * Output: [[-1,-1,2],[-1,0,1]]
 * Example 2:
 * <p>
 * Input: nums = []
 * Output: []
 * Example 3:
 * <p>
 * Input: nums = [0]
 * Output: []
 * <p>
 * <p>
 * Constraints:
 * <p>
 * 0 <= nums.length <= 3000
 * -10^5 <= nums[i] <= 10^5
 */
public class ThreeSum {

    public static void main(String[] args) {
        ThreeSum ts = new ThreeSum();
        System.out.println(ts.threeSum(new int[]{-1, 0, 1, 2, -1, -4}));
    }

    public List<List<Integer>> threeSum(int[] nums) {
        Arrays.sort(nums);
        List<List<Integer>> res = new ArrayList<>();

        for (int i = 0; i < nums.length - 2; i++) {
            if (nums[i] > 0) {
                break;
            }

            if (i > 0 && nums[i] == nums[i - 1]) {
                continue;
            }

            int left = i + 1;
            int right = nums.length - 1;

            while (right > left) {

                int sum = nums[i] + nums[left] + nums[right];

                if (sum > 0) {
                    --right;
                } else if (sum < 0) {
                    ++left;
                } else {
                    List<Integer> entry = new ArrayList<>();
                    entry.add(nums[i]);
                    entry.add(nums[left]);
                    entry.add(nums[right]);
                    res.add(entry);

                    while (right > left && nums[left] == nums[left + 1]) {
                        left++;
                    }

                    while (right > left && nums[right] == nums[right - 1]) {
                        right--;
                    }

                    ++left;
                    --right;
                }
            }
        }
        return res;
    }
}

File: src/main/java/com/atrilos/twoPointers/ThreeSum.kt
package com.atrilos.twoPointers

/**
 * [15](https://leetcode.com/problems/3sum/description/)
 */
fun threeSum(nums: IntArray): List<List<Int>> {
    val res = mutableListOf<List<Int>>()
    nums.sort()
    for (i in 0..nums.size - 3) {
        if (nums[i] > 0) break
        if (i > 0 && nums[i] == nums[i - 1]) continue
        var j = i + 1
        var k = nums.lastIndex
        while (j < k) {
            val sum = nums[i] + nums[j] + nums[k]
            if (sum == 0) {
                res.add(listOf(nums[i], nums[j], nums[k]))
                k--
                j++
                while (j < k && nums[j] == nums[j - 1]) j++
            } else if (sum > 0) {
                k--
            } else {
                j++
            }
        }
    }

    return res
}
